Binary number

energyc0 opened the pull request:
Binary numbers exercise is done.

energyc0 committed:
Binary numbers, allocate, create and print

energyc0 committed:
Binary numbers finished
Implemented addition, added comments, formatted code.

energyc0 committed:
Started adding tests, written core.

energyc0 committed:
Added binaryNumberCompare()

energyc0 committed:
binaryNumberConvertToStr()

energyc0 committed:
Changed named to binaryNumberToStr()

energyc0 committed:
Deleted macros, changed to functions, added binaryCheckAddtition() test.

energyc0 committed:
Added more tests.

WoWaster commented on 17.11.2025/binaryNumbers.h:
```diff
@@ -0,0 +1,69 @@
+ */
+char* binaryNumberToStr(const BinaryNumber* num, char* buf);
+
+#ifdef DEBUG
```
Как решите, но вообще в Си бывает стандартный `NDEBUG` имеющий обратное значение. 
https://en.cppreference.com/w/c/error/assert.html

WoWaster replied:
И прочитав PR целиком я честно не понял, зачем такие извращения. Хочется?)))

energyc0 replied:
> И прочитав PR целиком я честно не понял, зачем такие извращения. Хочется?)))

Ну, хочу отделить релиз от дебага.

WoWaster commented on 17.11.2025/binaryNumbers.h:
```diff
@@ -0,0 +1,69 @@
+ * Return 0 if they are equal or someone equals NULL.
+ * Return value > 0 if a > b, return value < 0 if a < b.
+ */
+int binaryNumberCompare(const BinaryNumber* a, const BinaryNumber* b);
```
Вроде не просили, но если хочется, то ок.

WoWaster commented on 17.11.2025/binaryNumbers.c:
```diff
@@ -0,0 +1,237 @@
+        return binNum;
+    }
+    for (int bit = BINARY_NUMBER_SIZE - 1; bit >= 0; bit--) {
+        binNum->data[bit] = ((decNum & (1 << (BINARY_NUMBER_SIZE - bit - 1))) ? 1 : 0);
```
По-моему, если двигать наоборот, будет сильно читаемее.
```suggestion
        binNum->data[bit] = (decNum >> (BINARY_NUMBER_SIZE - bit - 1)) & 1;
```
(Проверьте, что я на правильное количество бит двигаю :))

WoWaster commented on 17.11.2025/binaryNumbers.c:
```diff
@@ -0,0 +1,237 @@
+    for (int bit = BINARY_NUMBER_SIZE - 1; bit >= 0; bit--) {
+        int res = a->data[bit] + b->data[bit] + carry;
+        result->data[bit] = res % 2;
+        carry = res / 2;
```
Здесь получается не очень честно, потому что у Вас может получиться двойка. А если перейти на язык
электроники, то без этого можно обойтись :) https://en.wikipedia.org/wiki/Adder_(electronics)#Full_adder

energyc0 replied:
Буду учиться :100:

WoWaster commented on 17.11.2025/binaryRepresentation.c:
Было бы круто, если бы это ещё в CI запускалось, но к этому вернёмся в следующем семестре

WoWaster commented on 17.11.2025/binaryNumbers.c:
```diff
@@ -0,0 +1,237 @@
+
+void binaryCheckAddition(int line, const BinaryCheckAdditionArgs* args, int* returnCode)
+{
+    BinaryNumber* binA = binaryNumberCreate(args->a);
+    BinaryNumber* binB = binaryNumberCreate(args->b);
```
А чистить память? :(

energyc0 replied:
Убейте меня :(

WoWaster reviewed (changes requested):
Давайте сделаем чуть покрасивее)

energyc0 committed:
Refactored code.
Fixed memory leak in binaryCheckAddition().
Implemented full adder in binaryNumberAdd().

energyc0 committed:
Merge branch 'main' into binaryNumber
